### PR TITLE
jdbc dao implementations for all models are created

### PR DESCRIPTION
**Pull request by andromash**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Override
+    public List<Order> getAll() {
+        List<Order> orders = new ArrayList<>();
+        String query = "SELECT * FROM orders;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
+                orders.add(extractOrderFromResultSet(rs));
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: In method extractOrderFromResultSet() you are opening connection to DB inside TWR block with another opened connection - this directly contradicts to 11th paragraph of checklist.
The same situation in all other DAOs

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        }
+    }
+
+    private Order insertProducts(Order order) {
+        String queryToUpdateProducts = "INSERT INTO orders_products(order_id, product_id) "
+                + "VALUES(?,?);";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(queryToUpdateProducts)) {
+            for (int i = 0; i < order.getProducts().size(); i++) {
+                statement.setLong(1, order.getId());
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: We don't need to do this insert inside loop - value is the same for all products. Also IMHO for-each loop will look better here. 
This comment for other classes too.

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        String queryToDeleteProducts = "DELETE FROM orders_products WHERE order_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(queryToDeleteProducts)) {
+            statement.setLong(1, order.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't delete product from shopping cart - "
+                    + order, e);
+        }
+        return insertProducts(order);
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: I think in this method we should update both Products and userId, not only products

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:112]: I think in this method we should clear orders_products table for deleting order.
The same for delete() in other DAOs

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+@Dao
+public class ShoppingCartDaoJdbcImpl implements ShoppingCartDao {
+    @Override
+    public Optional<ShoppingCart> getByUser(Long userId) {
+        String query = "SELECT * FROM shopping_carts WHERE user_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, userId);
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]:
```suggestion
            if (rs.next()) {
```

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                    + id, e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public List<ShoppingCart> getAll() {
+        String query = "SELECT * FROM shopping_carts sc\n"
+                + "INNER JOIN shopping_carts_products scp ON sc.cart_id = scp.cart_id\n"
+                + "INNER JOIN products p ON scp.product_id = p.product_id\n";
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: Why do you need JOIN here? Below you don't use any fields from joined tables - only cart_id and user_id from shopping_carts table.

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        String queryToUpdateRoles = "INSERT INTO users_roles(user_id, role_id) "
+                + "VALUES(?,(SELECT role_id FROM roles WHERE role_name = ?)); ";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(queryToUpdateRoles)) {
+            for (Role role : user.getRoles()) {
+                statement.setLong(1, user.getId());
+                statement.setString(2, role.getRoleName().name());
+                statement.executeUpdate();
+            }
+            user.setRoles(getRolesOfUser(user.getId()));
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: What are you doing in this line?

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 2 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: Agree to the first part. But somehow foreach didn't work, the table was empty. Still, it works well with this loop

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 3 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: We can not change the user of order, it makes no sense

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 4 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:112]: But if we delete orders_products, what is the purpose of keeping orders in the table? they have no information except for id and user_id

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 5 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: I will stay with a while. It looks prettier for me :)

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 6 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: Agree.

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 7 above)
Comment (andromash) [reply src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: Till this moment users had roles, but their ids were null. So we take them from DB after inserting

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+@Dao
+public class OrderDaoJdbcImpl implements OrderDao {
+    @Override
+    public List<Order> getOrdersOfUser(Long userId) {
+        List<Order> orders = new ArrayList<>();
+        String query = "SELECT * FROM orders WHERE user_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, userId);
+            ResultSet rs = statement.executeQuery();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: rs - naming

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            while (rs.next()) {
+                orders.add(extractOrderFromResultSet(rs));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not get order from DB with user ID = "
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can not get orders from DB with user ID = "
```

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+            List<Product> products = new ArrayList<>();
+            while (resultSet.next()) {
+                Long productId = resultSet.getLong("product_id");
+                String name = resultSet.getString("name");
+                BigDecimal price = resultSet.getBigDecimal("price");
+                products.add(new Product(productId, name, price));
+            }
+            return products;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get shopping cart with ID = " + cartId, e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Couldn't get shopping cart products with ID = " + cartId, e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+
+    @Override
+    public Optional<ShoppingCart> get(Long id) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        String query = "SELECT * FROM shopping_carts sc\n"
+                + "INNER JOIN shopping_carts_products scp ON sc.cart_id = scp.cart_id\n"
+                + "INNER JOIN products p ON scp.product_id = p.product_id\n"
+                + "WHERE sc.cart_id = ?;";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: do we need info from products table, if we making another query in extractProductsFromShoppingCart method

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                    + id, e);
+        }
+        shoppingCart.setProducts(extractProductsFromShoppingCart(shoppingCart.getId()));
+        return Optional.of(shoppingCart);
+    }
+
+    @Override
+    public List<ShoppingCart> getAll() {
+        List<ShoppingCart> shoppingCarts = new ArrayList<>();
+        String query = "SELECT * FROM shopping_carts ";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: do we want to fetch deleted carts as well? 
not sure if we even need this method

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+import java.util.Set;
+
+@Dao
+public class UserDaoJdbcImpl implements UserDao {
+    @Override
+    public Optional<User> findByLogin(String login) {
+        User user = new User();
+        String query = "SELECT * FROM users u\n"
+                + "INNER JOIN users_roles ur ON u.user_id = ur.user_id\n"
+                + "INNER JOIN roles r ON ur.role_id = r.role_id\n"
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: do we need roles info? 
i guess we have separate method for it

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+            while (rs.next()) {
+                Long roleId = rs.getLong("role_id");
+                String roleName = rs.getString("role_name");
+                Role role = Role.of(roleName);
+                role.setId(roleId);
+                roles.add(role);
+            }
+            return roles;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get user with ID = " + userId, e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Couldn't get roles of user with ID = " + userId, e);
```

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.setLong(4, user.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't update user " + user, e);
+        }
+        try (Connection connection = ConnectionUtil.getConnection();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: i suggest we create separate private method for deleting roles

**Comment by RomanDubovskyi** [comment]
travis failed as well

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 4 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:112]: But at 11 paragraph of checklist we are suggested to make this deletion

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+    @Override
+    public Order update(Order order) {
+        String queryToDeleteProducts = "DELETE FROM orders_products WHERE order_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(queryToDeleteProducts)) {
+            statement.setLong(1, order.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't delete product from shopping cart - "
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: shopping cart?

**Review by RomanDubovskyi — APPROVED** [review]
fix exception message, rest seems to be good

**Pull request merged**